nn/transformer.py

```diff
@@ -0,0 +1,237 @@
+"""
+Transformer Modules
+"""
+
+from __future__ import annotations
+from typing import Optional, Any, Union, Callable
+from .. import nn
+
+
+class TransformerEncoderLayer(nn.Module):
+  """
+  Defines one layer of a standard transformer encoder
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, dropout: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu, layer_norm_eps: float = 1e-5,
+               norm_first: bool = False) -> None:
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param activation: activation function
+    :param layer_norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # to be implemented
+
+    self.linear_ff = nn.Linear(dim_ff)
+    self.linear_out = nn.Linear(dim_model)
+
+    self.activation = activation
+    self.norm_first = norm_first
+    self.norm_eps = layer_norm_eps
+    self.dropout = dropout
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    """
+    Two possible forward variants of encoder, defined by self.norm_first
+    """
+    x = inp
+    if self.norm_first:
+      x = x + self._self_attention_block(nn.layer_norm(x, epsilon=self.norm_eps))
+      x = x + self._feed_forward_block(nn.layer_norm(x, epsilon=self.norm_eps))
+    else:
+      x = nn.layer_norm(x + self._self_attention_block(x), epsilon=self.norm_eps)
+      x = nn.layer_norm(x + self._feed_forward_block(x), epsilon=self.norm_eps)
+
+    return x
+
+  def _self_attention_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.self_attn(x, x, x)
+    return nn.dropout(x, self.dropout)
+
+  def _feed_forward_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.linear_ff(x)
+    x = self.activation(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    x = self.linear_out(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    return x
+
+
+class TransformerEncoder(nn.Module):
+  """
+  Defines the full Encoder of the standard transformer
+  """
+  def __init__(self, encoder_layer: Union[TransformerEncoderLayer, Any], num_layers: int,
+               norm: Callable[[nn.LayerRef], nn.LayerRef] = None, layer_norm_eps: float = 1e-5):
+    """
+    :param encoder_layer: Encoder layer to be stacked num_layers times
+    :param num_layers: Number of layers
+    :param norm: normalization function
+    :param layer_norm_eps: Epsilon value for layer normalization
```

Comment:
Inconsistent to `norm`. Or is this specific to layer norm?

Reply:
It's still inconsistent.

Reply:
I changed it locally, marking it as done so I know what I have to still look at.

Reply:
No, push it first and then mark it as done.
But in any case, I guess you changed it to `norm_eps` now, right?

Reply:
Ah, ok. 
I changed the `layer_norm` prefix to `norm` everywhere it was used. So the normalization (and its parameters) now all have `norm` prefix.